### ISSUE TITLE
Let's zoom a brainstorm? JupyterLab keyboard navigation and usage

### ISSUE THREAD

**Issue by gabalafou** [jupyterlab/team-compass#210]
## Who?

Anyone interested! 

In particular, end-users who care about keyboard-related issues in JupyterLab (shortcuts, tab-navigation, etc.) and anyone who has worked on these issues (designers, developers, etc.).

## What?

A one-time Zoom meeting for people to join and brainstorm ideas around keyboard navigation in JupyterLab.

The meeting will be about generating ideas, not reaching decisions.

## When?

[August 30, 2023 at 17:00 UTC (click link for the time in your time zone)](https://www.timeanddate.com/worldclock/converter.html?iso=20230830T170000&p1=234&p2=179&p3=1440&p4=136&p5=195&p6=26)

90 minutes (but we don't have to use all 90 minutes)

## Where?

Zoom! [Zoom link in calendar event](https://calendar.google.com/calendar/r/eventedit/copy/NnN1MGZkNnQ2bm05bW9icm84Ym50OTJ0bTMgbTNoZWs2OWRhZzczODF1bXQ4a2NqZDc1dTRAZw)

## Why?

1. A number of PRs addressing various keyboard-navigation related issues have stalled in the `jupyterlab` repo,  such as but not limited to the following:
    - [Fix skiplink and add placeholder for additional skiplinks #14597](https://github.com/jupyterlab/jupyterlab/pull/14597)
    - [Fix tab trap notebook cells #14115](https://github.com/jupyterlab/jupyterlab/pull/14115)
    - [Activate every new current widget so that focus would not be reverted to 'body' while current widget exists #13921](https://github.com/jupyterlab/jupyterlab/pull/13921)
2. I believe that a brainstorming session could help unblock us.
3. @SylvainCorlay has mentioned [funding for keyboard-navigation related issues](https://github.com/jupyter/notebook/issues/6595#issuecomment-1604024622). A brainstorming session could be a useful input to that work.
4. I have been brainstorming with myself for weeks and weeks, and at this point, I really need other brains to storm with :). There's always too many other agenda items in other calls to have the time and space that I think we as a community could benefit from by more open-ended ideating around this topic.

I will provide a bit more context later. But I wanted to get this up before the end of the week!

## Notebook 7 or JupyterLab?

Both! I will provide more context later, but I want to encourage people to think about both Notebook 7 and JupyterLab when generating ideas to solve the keyboard navigation issues.

## Notification plan

1. Tagging people on this GitHub issue
2. Sharing it on Discourse
3. Sharing it in Gitter
4. Sharing it in the next JupyterLab team meeting

## Closing this issue

I will close this issue if any of the following have happened:

1. A time for the brainstorming session is scheduled
2. Someone asks me to move this issue somewhere else
3. A decision is reached to not move forward with this event

**Comment by gabalafou** [comment]
Hi! I'm tagging you on this issue because I have reason to believe that you might be interested (because of work on keyboard issues, or accessibility, or windowed rendering). I'm sure I missed someone, so please please please tag anyone else you think might be interested! 

cc @andrii-i, @t03857785, @isabela-pf, @trallard, @afshin, @fcollonval, @krassowski, @SylvainCorlay, @tonyfast, @ohrely, @marthacryan, @g547315, @vidartf, @scmmmh, @jasongrout, @GabrielaVives, @HaudinFlorence

**Comment by trallard** [comment]
I am interested! Let me know how I can be of help.

**Comment by andrii-i** [comment]
I am interested to participate as well. 

Also wanted to link outcomes of workshop on keyboard shortcuts hosted by @GabrielaVives and @HaudinFlorence: https://github.com/jupyterlab/jupyterlab/issues/14240.

**Comment by brichet** [comment]
Nice @gabalafou, I'm am also interested to participate.

**Comment by t03857785** [comment]
Definitely interested in joining this, if you agree a meeting I will do my best to join

**Comment by fcollonval** [comment]
I'll try to join too.

**Comment by gabalafou** [comment]
Hi friends, I'm so excited to see so much interest! The next step is to figure out a time:

1. :tada: August 23, [17:00 UTC](https://www.timeanddate.com/worldclock/converter.html?iso=20230816T170000&p1=234&p2=179&p3=1440&p4=136&p5=195&p6=26). This is immediately following the JupyterLab weekly team meeting time slot. 
2. :rocket: August 30, 17:00 UTC
3. :eyes: Other (please elaborate in a comment)

Please vote for your preferred time(s) using the emoji reactions. And if you think there's a better time, please don't hesitate to suggest it and explain why, thanks!

**Comment by HaudinFlorence** [comment]
I am interested to join with a preference for August 30th.

**Comment by gabalafou** [comment]
Okay, great! Let's do August 30th! 

I asked to block out 90 minutes following the JupyterLab team call, but we don't have to use all 90 minutes. 

[Mark your calendars! 😄](https://calendar.google.com/calendar/r/eventedit/copy/NnN1MGZkNnQ2bm05bW9icm84Ym50OTJ0bTMgbTNoZWs2OWRhZzczODF1bXQ4a2NqZDc1dTRAZw)

**Comment by GabrielaVives** [comment]
Thanks for setting this up! I'll join as well.

**Comment by gabalafou** [comment]
Closing issue because the meeting happened :)

**Comment by gabalafou** [comment]
Some of you might remember yesterday that there was a moment when I mentioned that [keyboard navigation for screen reader users](https://github.com/jupyterlab/jupyterlab/pull/15001#discussion_r1311452081) is probably a better story than keyboard navigation with just a keyboard and no screen reader.

I was going to try to screen-share demo what I meant, but it was a little too much for me at that moment to try to attend to the call while simultaneously running VoiceOver.

But this same topic came up in a pull request, and I left a comment there with a screenshot that explains what I was referring to yesterday in the call.

**Comment by tonyfast** [comment]
> Some of you might remember yesterday that there was a moment when I mentioned that [keyboard navigation for screen reader users](https://github.com/jupyterlab/jupyterlab/pull/15001#discussion_r1311452081) is probably a better story than keyboard navigation with just a keyboard and no screen reader.

yea, i was lost when this hasty comment was made. the screen reader navigation of lab and notebook is not better, it is more confusing and reveals more WCAG failures. the best we can say at this point is <q>there are landmarks</q>. they are not [assistive](https://github.com/Chartability/POUR-CAF#assistive). and, landmarks are further complicated by the fact that they have different meanings in lab and notebook.

![lab landmarks in orca](https://github.com/jupyterlab/team-compass/assets/4236275/a80ecf74-e6f8-495b-b2d5-008871ee0b48)

the figure above shows landmarks in lab with multiple notebooks open. the accNames for the regions are non-unique and failure WCAG. which notebooks cells? do i navigate to the toolbar directly? what can i interact with in the status bar that just called information? what is the main region? there is nothing interactive there. the repetitive, ambiguous naming should be considered for an assistive landmark navigation experience in lab. 

in notebook, the main region is apparent. we can't have landmarks to multiple documents, document nagivation is handled by tabs in the browser. below demonstrates how the notebooks for all project is hoping to test to the assistiveness of including cells as landmarks for screen reader users. 

![landmark navigation of single notebook documents that exposes cells as regions](https://github.com/jupyterlab/team-compass/assets/4236275/7b5b6ea4-61b1-48d4-a742-083db6820bf0)

further, we'd want to consider headings as another way to navigate the document. currently, lab with multiple notebooks open will fail WCAG because there are multiple h1s. Screen readers organize the headers by level and will make obscure navigation experiences in multi document mode. Headings, in single document, make A LOT of sense, and even [including code objects in the heading/landmark navigation](https://tonyfast.github.io/tonyfast/xxiii/2023-08-10-semantic-code-blocks.html#the-modified-highlighter-in-action) could improve the developer experience and embrace the literate programming qualities of the notebook.

![heading navigation in jupyterlab failing wcag h1 uniqueness](https://github.com/jupyterlab/team-compass/assets/4236275/e520c3af-50be-4dd0-b6c1-1789c14d1d59)

a focused event on landmarks, heading, and screen reader navigation might be a good workshop. [accessibility bookmarklets](https://accessibility-bookmarklets.org/) could allow non screen reader users to understand these navigation pattern. this event would spend a lot of time considering the roles of elements, their accName, and their accDesc. understanding these relationships will improve the quality of the jupyter experience for everyone.
